src/tools/openpgp-tool.c

```diff
@@ -441,29 +664,62 @@ static int do_dump_do(sc_card_t *card, unsigned int tag)
 	return EXIT_SUCCESS;
 }
 
-int do_genkey(sc_card_t *card, u8 key_id, unsigned int key_len)
+int do_genkey(sc_card_t *card, u8 key_id, const char *keytype)
 {
 	int r;
 	sc_cardctl_openpgp_keygen_info_t key_info;
 	u8 fingerprints[60];
 	sc_path_t path;
 	sc_file_t *file;
 
+	/* validate key_id */
 	if (key_id < 1 || key_id > 3) {
 		util_error("unknown key ID %d", key_id);
 		return SC_ERROR_INVALID_ARGUMENTS;
 	}
+
+	/* fack back to RSA 2048 if keytype is not given */
+	if (!keytype)
+		keytype = "RSA2048";
+
 	memset(&key_info, 0, sizeof(sc_cardctl_openpgp_keygen_info_t));
-	key_info.key_id = key_id;
-	key_info.algorithm = SC_OPENPGP_KEYALGO_RSA;
-	key_info.rsa.modulus_len = key_len;
-	key_info.rsa.modulus = malloc(key_len/8);
-	r = sc_card_ctl(card, SC_CARDCTL_OPENPGP_GENERATE_KEY, &key_info);
-	free(key_info.rsa.modulus);
-	if (r < 0) {
-		util_error("failed to generate key: %s", sc_strerror(r));
+
+	/* generate key depending on keytype passed */
+	if (strncasecmp("RSA", keytype, strlen("RSA")) == 0) {
```

Comment:
Using `sscanf()` would be easier, wouldn't it?

Reply:
Not really.
I am trying to parse a regular expression `/rsa(\d+)?/i`, i.e. a string
1) starting with the literal string "RSA" in any upper-/lower-case combination
2) optionally followed by a positive number, which terminates the string

Unfortunately `scanf ("%u",...)` has some strange effects when being passed negative numbers.
With all the checks to cover that plus the checks for the correct size, the difference in LoC is negligible.